src/all/hdoujin/src/eu/kanade/tachiyomi/extension/all/hdoujin/HDoujin.kt

```diff
@@ -0,0 +1,397 @@
+    private fun alwaysIncludeTags() = preferences.getString(PREF_INCLUDE_TAGS, "")
+    private fun alwaysExcludeTags() = preferences.getString(PREF_EXCLUDE_TAGS, "")
+    private fun getTagsPreference(): String {
+        val include = alwaysIncludeTags()
+            ?.split(",")
+            ?.map(String::trim)
+            ?.filter(String::isNotBlank)
+
+        val exclude = alwaysExcludeTags()
+            ?.split(",")
+            ?.map(String::trim)
+            ?.filter(String::isNotBlank)
+            ?.map { "-$it" }
+
+        val tags: List<String> = include?.plus(exclude ?: emptyList()) ?: exclude?.plus(include ?: emptyList()) ?: emptyList()
+        if (tags.isNotEmpty()) {
+            val tagGroups: Map<String, Set<String>> = tags
+                .groupBy {
+                    val tag = it.removePrefix("-")
+                    val parts = tag.split(":", limit = 2)
+                    if (parts.size == 2 && parts[0].isNotBlank()) parts[0] else "tag"
+                }
+                .mapValues { (_, values) ->
+                    values.mapTo(mutableSetOf()) {
+                        val tag = it.removePrefix("-").split(":").last().trim()
+                        if (it.startsWith("-")) "-$tag" else tag
+                    }
+                }
+
+            return tagGroups.entries.joinToString(" ") { (key, values) ->
+                "$key:\"${values.joinToString(",")}\""
+            }
+        }
+        return ""
+    }
+
+    override val baseUrl: String = "https://hdoujin.org"
+    private val baseApiUrl: String = "https://api.hdoujin.org"
+    private val bookApiUrl: String = "$baseApiUrl/books"
+
+    override fun headersBuilder() = super.headersBuilder()
+        .set("Referer", "$baseUrl/")
+        .set("Origin", baseUrl)
+
+    private val context: Application by injectLazy()
+    private val handler by lazy { Handler(Looper.getMainLooper()) }
+    private var _clearance: String? = null
+
+    @SuppressLint("SetJavaScriptEnabled")
+    fun getClearance(): String? {
+        _clearance?.also { return it }
+        val latch = CountDownLatch(1)
+        handler.post {
+            val webview = WebView(context)
+            with(webview.settings) {
+                javaScriptEnabled = true
+                domStorageEnabled = true
+                databaseEnabled = true
+                blockNetworkImage = true
+            }
+            webview.webViewClient = object : WebViewClient() {
+                override fun onPageFinished(view: WebView?, url: String?) {
+                    view!!.evaluateJavascript("window.localStorage.getItem('clearance')") { clearance ->
+                        webview.stopLoading()
+                        webview.destroy()
+                        _clearance = clearance.takeUnless { it == "null" }?.removeSurrounding("\"")
+                        latch.countDown()
+                    }
+                }
+            }
+            webview.loadUrl("$baseUrl/")
```

Comment:
```suggestion
            webview.loadDataWithBaseURL("$baseUrl/", " ", "text/html", null, null)
```
we don't actually need to load the website, so prevent one network call